upgrade-takeover.sh

```diff
@@ -0,0 +1,460 @@
+            continue
+        fi
+        log WARN "Retrying failure report - try $c"
+        sleep 60
+    done
+}
+
+# Log operational message; writes provided text to journal and echos to stdout.
+# If log at ERROR level, exits this script with code 1.
+#
+# $1 -- optional log level, must be ERROR or WARN; otherwise defaults to INFO
+# $2 -- log message
+function log {
+    # Process log level if provided to function
+    priority=6
+    case $1 in
+        ERROR)
+            loglevel=ERROR
+            # 3 is "err"
+            priority=3
+            shift
+            ;;
+        WARN)
+            loglevel=WARNING
+            # 4 is "warning"
+            priority=4
+            shift
+            ;;
+        *)
+            loglevel=INFO
+            ;;
+    esac
+    echo "${1}" | systemd-cat --level-prefix=0 --identifier="${SCRIPTNAME}" --priority="${priority}" 2> /dev/null || true
+    endtime=$(date +%s)
+    printf "[%s][%09d%s%s\n" "$SCRIPTNAME" "$((endtime - starttime))" "][$loglevel]" "$1"
+    if [ "$loglevel" == "ERROR" ]; then
+        exit 1
+    fi
+}
+
+# Test if the first arg is greater than the second, when the args are compared
+# as semvers.
+# Return 0 if the first item is greater; 1 otherwise
+# For example, version_gt "1.2.10" "1.2.3" returns true.
+#
+# $1 -- expected greater version
+# $2 -- version to compare
+function version_gt() {
+    test "$(echo "$@" | tr " " "\n" | sort -V | head -n 1)" != "$1"
+}
+
+# Compare the provided percentage and state values with backend device API
+# provisioning_progress and provisioning_state.
+#
+# $1 -- local percentage
+# $2 -- local state
+#
+# Return 0 if the values are equal; otherwise return 1
+function compare_device_state() {
+    pct=$1
+    state=$2
+    resp=$(CURL_CA_BUNDLE="${TMPCRT}" ${CURL} --header "Authorization: Bearer ${APIKEY}" \
+        "${API_ENDPOINT}/v6/device(uuid='${UUID}')?\$select=provisioning_state,provisioning_progress" | jq '.d[]')
+    remote_pct=$(echo "${resp}" | jq -r '.provisioning_progress')
+    remote_state=$(echo "${resp}" | jq -r '.provisioning_state')
+    if [ -n "${remote_pct}" ] && [ -n "${remote_state}" ]; then
+        test "${pct}" -eq "${remote_pct}" && test "${state}" = "${remote_state}"
+    else
+        return 1
+    fi
+}
+
+# Stop Supervisor and related systemd services
+function stop_services() {
+    log "Stopping supervisor and related services..."
+    systemctl stop update-balena-supervisor.timer > /dev/null 2>&1 || systemctl stop update-resin-supervisor.timer > /dev/null 2>&1
+    systemctl stop balena-supervisor  > /dev/null 2>&1 || systemctl stop resin-supervisor > /dev/null 2>&1
+    ${DOCKER_CMD} rm -f balena_supervisor resin_supervisor > /dev/null 2>&1 || true
+}
+
+# Remove and log fsck recovery files. Since takeover will reflash the disk, it's
+# not strictly necessary. However, it may indicate issues with the storage medium,
+# so good to know.
+function remove_rec_files() {
```

Comment:
How does `fsck` help in this case if takeover will overwrite all of the partitions?

Reply:
I was following up on the comment above - `fsck` or similar would show whether the media is healthy before copying and launching the takeover tool. How takeover would fail from a damaged storage is difficult to predict - probably failing to write and handling the error, but it's probably an untested path.

However, if it fails to write because of damaged sectors, it could have already overwritten parts so the device will be bricked. Bailing out if fsck errors is a way to improve reliability.

Reply:
Sounds fine to me. So to be clear, you're saying use `fsck -n`, capture stdout+stderr, and failure is defined as any non-zero result. Log the output on failure. Use `lsblk` to find the device name of the `/mnt/data` partition?

Let me know of any better ways; not my expertise.

Reply:
Yeah, something like:
```
fsck.ext4 -n /dev/sda1 > dev/null
if [ $? -ge 4 ]; then
    exit 1
fi
```

Reply:
Done; to be clear, only checking data partition.